airbyte-migration/src/main/java/io/airbyte/migrate/migrations/MigrationV0_18_0.java

```diff
@@ -0,0 +1,191 @@
+import com.google.common.collect.ImmutableMap;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.commons.util.MoreIterators;
+import io.airbyte.migrate.Migration;
+import io.airbyte.migrate.ResourceId;
+import io.airbyte.migrate.ResourceType;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This migration makes sure that ConfiguredAirbyteCatalog always have values for the now required
+ * fields: syncMode (used by source to specify full_refresh/incremental) and destinationSyncMode
+ * (used by destination to specify append/overwrite/append_dedup)
+ *
+ * The primaryKey column is filled if available from the stream if defined by source
+ */
+public class MigrationV0_18_0 extends BaseMigration implements Migration {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MigrationV0_18_0.class);
+
+  private static final ResourceId STANDARD_SYNC_RESOURCE_ID = ResourceId.fromConstantCase(ResourceType.CONFIG, "STANDARD_SYNC");
+
+  private static final String MIGRATION_VERSION = "0.18.0-alpha";
+
+  private final Migration previousMigration;
+
+  public MigrationV0_18_0(Migration previousMigration) {
+    super(previousMigration);
+    this.previousMigration = previousMigration;
+  }
+
+  @Override
+  public String getVersion() {
+    return MIGRATION_VERSION;
+  }
+
+  @Override
+  public Map<ResourceId, JsonNode> getOutputSchema() {
+    return previousMigration.getOutputSchema();
+  }
+
+  @Override
+  public void migrate(Map<ResourceId, Stream<JsonNode>> inputData, Map<ResourceId, Consumer<JsonNode>> outputData) {
+    for (final Map.Entry<ResourceId, Stream<JsonNode>> entry : inputData.entrySet()) {
+      final Consumer<JsonNode> recordConsumer = outputData.get(entry.getKey());
+
+      entry.getValue().forEach(r -> {
+        if (entry.getKey().equals(STANDARD_SYNC_RESOURCE_ID)) {
+          ((ObjectNode) r).set("catalog", migrateCatalog(r.get("catalog")));
+        }
+        recordConsumer.accept(r);
+      });
+    }
+  }
+
+  private JsonNode migrateCatalog(JsonNode catalog) {
+    final List<Map<String, JsonNode>> configuredStreams = MoreIterators.toList(catalog.get("streams").elements())
+        .stream()
+        .map(stream -> {
+          final JsonNode airbyteStream = stream.get("stream");
+          JsonNode syncMode = stream.get("sync_mode");
+          if (syncMode == null) {
+            syncMode = Jsons.jsonNode(SyncMode.FULL_REFRESH.toString());
+            LOGGER.info("Migrating {} to default source sync_mode: {}", airbyteStream.get("name"), syncMode);
+          }
+          JsonNode destinationSyncMode = stream.get("destination_sync_mode");
+          if (destinationSyncMode == null) {
+            if (syncMode != null && SyncMode.fromValue(syncMode.asText()) == SyncMode.FULL_REFRESH) {
```

Comment:
shouldn't syncMode always be not null because of what you did on lines 96-99?

Reply:
i think because of that you can get rid of the else branch and stop checking `syncMode != null` in the other branches